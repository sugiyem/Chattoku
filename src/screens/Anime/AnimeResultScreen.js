import React, { useEffect, useState } from "react";
import { Button, ButtonText, Container } from "../../styles/GeneralStyles";

import AnimeCollection from "../../components/Anime/AnimeCollection";
import AnimeFetch from "../../services/Anime/AnimeFetch";
import AnimeSearchBar from "../../components/Anime/AnimeSearchBar";
import FetchFavoriteAnime from "../../services/Anime/FetchFavoriteAnime";
import { fetchType } from "../../constants/MyAnimeList";
import { ActivityIndicator } from "react-native";

const AnimeResultScreen = ({ navigation, route }) => {
  const [search, setSearch] = useState(route.params.search);
  const [animeData, setAnimeData] = useState([]);
  const [resultExpanded, setResultExpanded] = useState(null);
  const [resultPage, setResultPage] = useState(1);
  const [favoriteList, setFavoriteList] = useState([]);
  const [isResultLoaded, setIsResultLoaded] = useState(false);

  const animeCollectionItems = [
    {
      title: "Search Result",
      data: animeData,
      expand: resultExpanded,
      changeExpand: setResultExpanded,
      page: resultPage,
      changePage: setResultPage
    }
  ];

  useEffect(() => {
    return FetchFavoriteAnime({
      onSuccesfulFetch: (favorite) => {
        setFavoriteList(favorite.map((item) => item.id));
      },
      onFailure: (error) => {
        Alert.alert(error.message);
      }
    });
  }, []);

  useEffect(() => {
    const abortController = new AbortController();
    setIsResultLoaded(false);

    AnimeFetch({
      type: fetchType.SEARCH,
      page: resultPage,
      search: search,
      onSuccesfulFetch: (data) => {
        setAnimeData(data);
        setIsResultLoaded(true);
      },
      abortController: abortController
    });

    return () => abortController.abort();
  }, [resultPage]);

  if (!isResultLoaded) {
    return (
      <Container>
        <ActivityIndicator size="large" />
      </Container>
    );
  }

  return (
    <Container>
      <AnimeSearchBar
        value={search}
        onChangeText={setSearch}
        navigation={navigation}
      />

      <Button onPress={() => navigation.goBack()}>
        <ButtonText>Go back</ButtonText>
      </Button>

      <AnimeCollection items={animeCollectionItems} favorite={favoriteList} />
    </Container>
  );
};

export default AnimeResultScreen;
