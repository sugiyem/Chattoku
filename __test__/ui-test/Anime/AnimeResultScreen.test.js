import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react-native";
import AnimeResultScreen from "../../../src/screens/Anime/AnimeResultScreen";

const mockGoBack = jest.fn();
const mockSetState = jest.fn();
const mockNavigation = {
  goBack: mockGoBack
};
const mockRoute = {
  params: {
    search: "test"
  }
};

jest.mock("@react-navigation/core", () => ({
  useNavigation: () => mockNavigation
}));

jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");
jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: (initial) => [
    typeof initial === "boolean" ? true : initial,
    mockSetState
  ],
  useEffect: () => {}
}));

describe("Test AnimeHomeScreen UI", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <AnimeResultScreen navigation={mockNavigation} route={mockRoute} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Navigation button is working", () => {
    const { getByTestId } = render(
      <AnimeResultScreen navigation={mockNavigation} route={mockRoute} />
    );

    fireEvent.press(getByTestId("backButton"));

    expect(mockGoBack).toHaveBeenCalled();
  });

  it("Search value is correct", () => {
    const { getByTestId } = render(
      <AnimeResultScreen navigation={mockNavigation} route={mockRoute} />
    );

    expect(getByTestId("searchInput").props.value).toEqual(
      mockRoute.params.search
    );
  });
});
