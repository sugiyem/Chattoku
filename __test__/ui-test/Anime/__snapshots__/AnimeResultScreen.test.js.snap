// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AnimeHomeScreen UI renders correctly 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  proxiedProperties={
    Object {
      "colors": Array [
        4287090411,
        4283256319,
      ],
      "endPoint": Array [
        1,
        1,
      ],
      "locations": undefined,
      "startPoint": Array [
        0,
        0,
      ],
    }
  }
  style={
    Object {
      "flex": 1,
      "padding": 5,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      onChangeText={[MockFunction]}
      placeholder="Search anime by original title"
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "borderColor": "#000000",
            "borderRadius": 10,
            "borderWidth": 1,
            "color": "#000000",
            "flexBasis": 0,
            "flexGrow": 3,
            "flexShrink": 1,
            "marginBottom": 5,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 5,
            "paddingBottom": 2,
            "paddingLeft": 2,
            "paddingRight": 2,
            "paddingTop": 2,
          },
        ]
      }
      testID="searchInput"
      value="test"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#483d8b",
          "borderRadius": 10,
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginBottom": 5,
          "marginLeft": 5,
          "marginRight": 5,
          "marginTop": 5,
          "opacity": 1,
          "paddingBottom": 5,
          "paddingLeft": 5,
          "paddingRight": 5,
          "paddingTop": 5,
        }
      }
      testID="searchButton"
    >
      <Text
        color="#ffffff"
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 14,
              "textAlign": "center",
            },
          ]
        }
      >
        Search
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "#7FFFD4",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginBottom": 5,
        "marginLeft": 5,
        "marginRight": 5,
        "marginTop": 5,
        "opacity": 1,
        "paddingBottom": 5,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 5,
      }
    }
    testID="backButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#483d8b",
            "fontSize": 14,
            "textAlign": "center",
          },
        ]
      }
    >
      Go back
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "changeExpand": [MockFunction],
          "changePage": [MockFunction],
          "data": Array [],
          "expand": null,
          "page": 1,
          "title": "Search Result",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    maxToRenderPerBatch={60}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
      ]
    }
    style={
      Object {
        "alignSelf": "stretch",
        "margin": 5,
        "paddingHorizontal": 5,
      }
    }
    windowSize={30}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Gill Sans",
                "fontSize": 30,
                "fontWeight": "bold",
                "textAlign": "center",
                "textDecorationLine": "underline",
              },
              Object {
                "marginBottom": 20,
              },
            ]
          }
          underline={true}
        >
          Search Result
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 5,
            }
          }
        >
          <View />
          <View />
        </View>
      </View>
    </View>
  </RCTScrollView>
</ViewManagerAdapter_ExpoLinearGradient>
`;
